### DISCUSSION TITLE
Proxy cookie handling

### DISCUSSION THREAD

**Discussion by ejheil** [narrowlink/narrowlink#95] [Q&A]
Hi - I don't have enough information to make this a useful bug report, but I thought I'd ask for ideas.  I'm trying to proxy the FreshRSS feed reader through a narrowlink web gateway, and it's telling me " The web server seems to be incorrectly configured for cookies required for PHP sessions!"

Digging around the FreshRSS github I didn't come up with a simple explanation of exactly what's missing, and honestly I'll be fine if it turns out I can't run FreshRSS through narrowlink just due to its PHP crankiness. 

 But I thought I'd ask in case there is some configuration in narrowlink I can tweak that will affect how its proxying affects cookies.

**Comment by ejheil** [comment]
(I should mention: I have other apps running just fine through narrowlink; FreshRSS is the first one I've had any troubles with.)

**Comment by SajjadPourali** [comment]
Hello,

It is crucial for me to address the potential bugs on Narrowlink, so let's conduct an investigation.

There is no manipulation or modification of cookies in Narrowlink, and they should transmit as they are.

I have two hypotheses that might affect FreshRSS:

1. NL supports HTTP/2 only over HTTPS. If you have set up HTTPS, your browser initiates a connection using HTTP/2, then the gateway converts the HTTP/2 request to HTTP/1 and sends it to your original server. There is a very small chance that the cookies could be affected during this transition. You can temporarily disable HTTP/2 from about:config in your browser and try again.

2. You may be using the HTTP protocol on Narrowlink, while the PHP cookies come with the 'secure' flag only. This causes the browser not to send the cookies to the server.

Please verify that the application works correctly on the local machine currently (using the original domain name for checking, you can edit the /etc/hosts file for this purpose). Also, verify that the Domain and Path attributes in the set-cookie response align with your configuration.

If the above information doesn't help, you can compare the HTTP requests and responses that you see over Narrowlink with the ones you see in direct access. You can use developer options or tools like mitmproxy/burpsuite to access your communications.

If you have any further questions or need assistance, please feel free to ask.

------

P.S. I am rewriting the main components of Narrowlink to improve performance(using protobuf, multiplexing overwebsocket), scalability(horizontal), and simplicity(web panel) for the future https://github.com/narrowlink/narrowlink/pull/92. I'll also revisit this part to ensure everything works well.

**Comment by ejheil** [comment]
Switching off http/2 in firefox allowed me to log in immediately!

**Reply by SajjadPourali** [reply]
I could reproduce the issue; let me conduct further investigation on it.

**Reply by ejheil** [reply]
Sounds great!

**Reply by SajjadPourali** [reply]
The problem seems to be related to "x-forwarded-xxx." It appears that I didn't handle it properly in narrowlink. If I've recognized the issue correctly, I expect to release the patch very soon.

**Comment by SajjadPourali** [comment]
I've just patched the issue https://github.com/narrowlink/narrowlink/pull/97 . You can refer to https://datatracker.ietf.org/doc/html/rfc9113#name-compressing-the-cookie-head to learn more about it.

Thank you for reporting this issue.

**Comment by SajjadPourali** [comment]
I just rolled out version 0.2.4, which addressed this issue https://github.com/narrowlink/narrowlink/releases/tag/0.2.4

**Reply by ejheil** [reply]
Confirmed, this fixes my cookie problem and I can get FreshRSS working!
